extras/sht3x/sht3x.c

```diff
@@ -0,0 +1,403 @@
+ * modification, are permitted provided that the following conditions are met:
+ * 
+ * 1. Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution.
+ *
+ * 3. Neither the name of the copyright holder nor the names of its 
+ * contributors may be used to endorse or promote products derived from this
+ * software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * Driver for Sensirion SHT3x digital temperature and humity sensor
+ * connected to I2C
+ *
+ * Part of esp-open-rtos
+ */
+
+#include <string.h>
+
+#include "sht3x.h"
+
+#include "FreeRTOS.h"
+#include "task.h"
+
+#include "espressif/esp_common.h"
+#include "espressif/sdk_private.h"
+
+#define SHT3x_STATUS_CMD               0xF32D
+#define SHT3x_CLEAR_STATUS_CMD         0x3041
+#define SHT3x_RESET_CMD                0x30A2
+#define SHT3x_FETCH_DATA_CMD           0xE000
+#define SHT3x_HEATER_OFF_CMD           0x3066
+
+const uint16_t SHT3x_MEASURE_CMD[6][3] = { {0x2c06,0x2c0d,0x2c10},    // [SINGLE_SHOT][H,M,L]
+                                           {0x2032,0x2024,0x202f},    // [PERIODIC_05][H,M,L]
+                                           {0x2130,0x2126,0x212d},    // [PERIODIC_05][H,M,L]
+                                           {0x2236,0x2220,0x222b},    // [PERIODIC_05][H,M,L]
+                                           {0x2234,0x2322,0x2329},    // [PERIODIC_05][H,M,L]
+                                           {0x2737,0x2721,0x272a} };  // [PERIODIC_05][H,M,L]
+
+#define max(a,b)    (a) > (b) ? a : b
+
+// tick counts [High, Medium, Low]
+// minimum is one tick if portTICK_PERIOD_MS is greater than 20 or 30 ms
+const int32_t SHT3x_MEASURE_DURATION[3] = { max(30/portTICK_PERIOD_MS,1),
+                                            max(20/portTICK_PERIOD_MS,1),
+                                            max(20/portTICK_PERIOD_MS,1) };
+
```

Comment:
This might still round down to a period that is too short. What about the following and then remove the `max` macro.

```
#define TIME_TO_TICK_CEILING(ms) (((ms) + (portTICK_PERIOD_MS - 1)) / portTICK_PERIOD_MS)
```

Reply:
Sorry, the rtos tick period can be short in rare cases, on a cpu freq change it can halve for one tick, so perhaps this should be:

```
#define TIME_TO_TICKS_CEILING(ms) (((ms) + (portTICK_PERIOD_MS / 2) + (portTICK_PERIOD_MS - 1)) / portTICK_PERIOD_MS)
```

Reply:
Hm, quite complicated.

I had supposed that your first formula should work well because it simply moves all values that are not divisible by portTICK_PERIOD_MS to a value that is higher than the next divisible value. For example, for a portTICK_PERIOD_MS of 10 ms, a duration of 32 ms is shifted to 41 ms and results in 4 ticks, while 30 ms are shifted to 39 ms and 3 ticks are obtained.

Their second formula leads to tick values that are too high for half of the values in normal cases. For example, for a portTICK_PERIOD_MS of 10 ms, a duration of 30 ms gives 44 ms and results in 4 ticks.

Since the durations given in the lookup table are too high, I would use your first formula. For example, according to the data sheet, a measurement with high repeatability (first value in the table) takes 13.5 ms. When I used 20 ms as the next multiple of portTICK_PERIOD_MS, I observed sometimes errors because the resulting waiting time was too short.